vmm/src/device_manager.rs

```
@@ -1325,88 +1328,91 @@ impl DeviceManager {
 
-                    devices.push((
-                        Arc::clone(&virtio_fs_device) as Arc<Mutex<dyn vm_virtio::VirtioDevice>>,
+                self.memory_manager
+                    .lock()
+                    .unwrap()
+                    .create_userspace_mapping(
+                        fs_guest_addr.raw_value(),
+                        fs_cache,
+                        addr,
+                        false,
                         false,
-                    ));
+                    )
+                    .map_err(DeviceManagerError::MemoryManager)?;
+
+                let mut region_list = Vec::new();
+                region_list.push(VirtioSharedMemory {
+                    offset: 0,
+                    len: fs_cache,
+                });
+
+                Some((
+                    VirtioSharedMemoryList {
+                        addr: fs_guest_addr,
+                        len: fs_cache as GuestUsize,
+                        region_list,
+                    },
+                    addr,
+                ))
+            } else {
+                None
+            };
+
+            let virtio_fs_device = Arc::new(Mutex::new(
+                vm_virtio::vhost_user::Fs::new(
+                    fs_sock,
+                    &fs_cfg.tag,
+                    fs_cfg.num_queues,
+                    fs_cfg.queue_size,
+                    cache,
+                )
+                .map_err(DeviceManagerError::CreateVirtioFs)?,
+            ));
+
+            self.add_migratable_device(Arc::clone(&virtio_fs_device) as Arc<Mutex<dyn Migratable>>);
+
+            return Ok((Arc::clone(&virtio_fs_device) as VirtioDeviceArc, false));
+        }
 
-                    let migratable = Arc::clone(&virtio_fs_device) as Arc<Mutex<dyn Migratable>>;
-                    let id = migratable.lock().unwrap().id();
-                    self.migratable_devices.insert(id, migratable);
+        Err(DeviceManagerError::NoVirtioFsSock)
```

Thanks for adding this error, this would prevent the code from silently ignoring malformed `--fs` parameters.